src/vision/status_quo/barbara_benchmarks_async_trait.md

```diff
@@ -0,0 +1,134 @@
+
+Barbara thinks about writing down the name of each Future type, but realizes that this wouldn't work with the `async` keyword. Using Future combinators directly would be extremely verbose and painful.
+
+Barbara finds the `async_trait` crate. Given her performance constraints, she is wary of the allocations and dynamic dispatch introduced by the crate.
+
+She decides to write a benchmark to simulate the performance impact of `async_trait` compared to a future where `async fn` is fully supported in traits. Looking at the `async_trait` documentation, she sees that it desugars code like
+
+```rust
+#[async_trait]
+impl Trait for Foo {
+    async fn run(&self) {
+        // ...
+    }
+}
+```
+
+to
+
+```rust
+impl Trait for Foo {
+    fn run<'a>(
+        &'a self,
+    ) -> Pin<Box<dyn std::future::Future<Output = ()> + Send + 'a>>
+    where
+        Self: Sync + 'a,
+    {
+        async fn run(_self: &Foo) {
+            // original body
+        }
+        Box::pin(run(self))
+    }
+}
+```
+
+The benchmark Barbara uses constructs a tree of Futures 5 levels deep, using both `async` blocks and a manual desugaring similar to above. She runs the benchmark on hardware that is representative for her use case and finds that while executing a single native async future takes 639 ns, the manual desugaring using `boxed` takes 1.82 µs.
+
+Barbara sees that in a real codebase, this performance would not be good enough for writing a network stack capable of sustaining gigabit-level throughput on underpowered hardware. Barbara is disappointed, but knows that support for `async fn` in traits is in the works.
+
+Barbara looks at her organization's current priorities and decides that 100's of mbps will be an acceptable level of performance for the near term. She decides to adopt `async_trait` with the expectation that the performance benefit will go away in the long term.
+
+## 🤔 Frequently Asked Questions
+
+*Here are some standard FAQ to get you started. Feel free to add more!*
+
+### **What are the morals of the story?**
+*Talk about the major takeaways-- what do you see as the biggest problems.*
+
+* Language features that don't work well together can be a major roadblock in the course of development. Developers expect all of a language's features to be at their disposal, not using one to cut them off from using another.
+* Allocation and dynamic dispatch aren't acceptable runtime performance costs for all use cases.
+
+### **What are the sources for this story?**
+*Talk about what the story is based on, ideally with links to blog posts, tweets, or other evidence.*
+
+This story is based on actual experience implementing the 3rd-generation network stack for the Fuchsia operating system.
+
+The benchmarks are [implemented here](https://fxrev.dev/528302).
+
+### **Why do you need to model backpressure?**
+
+The Linux network stack doesn't do this; instead it drops packets as hardware buffers fill up.
```

Comment:
Well, not exactly. The TCP stack and associated system calls certainly exercise backpressure. Only UDP and packet based APIs are not.  Don't think this changes the story a lot however ;)